### PR TITLE
Mobile Recaptcha

### PR DESCRIPTION
**Pull request by cristhianescobar**
Android and iOS Recaptcha Implementation

### CODE REVIEW

File: shared/src/commonMain/kotlin/io.newm.shared/config/NewmSharedBuildConfig.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    val walletConnectProjectId: String
+        get() = SharedBuildConfig.WALLET_CONNECT_PROJECT_ID
+
+    val recaptchaSiteKey: String
+        get() = SharedBuildConfig.RECAPTCHA_SITE_KEY
+
+    val isDevMode: Boolean
+        get() = mode == Mode.DEVELOPMENT
+}
```
Comment (martyu) [inline shared/src/commonMain/kotlin/io.newm.shared/config/NewmSharedBuildConfig.kt]: Why are these in kmm if they're not used in kmm?  (besides walletConnectProjectId)

Comment (martyu) [reply shared/src/commonMain/kotlin/io.newm.shared/config/NewmSharedBuildConfig.kt]: iOS needs some of these at compile time, but this object is only available at runtime.

Comment (cristhianescobar) [reply shared/src/commonMain/kotlin/io.newm.shared/config/NewmSharedBuildConfig.kt]: I am exposing them through KMP, so both clients can have access to it, these values are available at compile time, there is code that gets generated before the app actually runs.

File: android/app-newm/src/main/java/io/newm/LoginActivity.kt [hunk 2]
```diff
@@ -112,6 +121,20 @@ class LoginActivity : ComponentActivity() {
...
+        lifecycleScope.launch {
+            Recaptcha.getClient(application, NewmSharedBuildConfig.recaptchaSiteKey, timeout = 20000L)
+                .onSuccess { client ->
+                    recaptchaClientProvider.setRecaptchaClient(client)
+                }
+                .onFailure { exception ->
+                    // Handle communication errors ...
+                    // See "Handle communication errors" section
+                    Log.e("LoginActivity", "Setup failed", exception)
+                }
```
Comment (newmskywalker) [inline android/app-newm/src/main/java/io/newm/LoginActivity.kt:135]: should we just crash? Or can the app function without this?

File: android/features/login/src/main/java/io/newm/feature/login/screen/createaccount/CreateAccountScreenPresenter.kt [hunk 3]
```diff
@@ -65,9 +69,14 @@ class CreateAccountScreenPresenter constructor(
...
-                                    signupUseCase.requestEmailConfirmationCode(
-                                        email = userEmail.text,
-                                    )
+                                    recaptchaClientProvider.get().execute(RecaptchaAction.SIGNUP).onSuccess { token ->
+                                        signupUseCase.requestEmailConfirmationCode(
+                                            email = userEmail.text,
+                                            humanVerificationCode = token
+                                        )
+                                    }.onFailure {
+                                        errorMessage = "Are you even a human?"
```
Comment (newmskywalker) [inline android/features/login/src/main/java/io/newm/feature/login/screen/createaccount/CreateAccountScreenPresenter.kt:78]: should we reset the state to `EmailAndPassword` on falure?

File: android/features/login/src/main/java/io/newm/feature/login/screen/createaccount/CreateAccountScreenPresenter.kt [hunk 4]
```diff
@@ -94,15 +103,19 @@ class CreateAccountScreenPresenter constructor(
...
+                                            email = userEmail.text,
+                                            verificationCode = verificationCode.text,
+                                            password = password.text,
+                                            passwordConfirmation = passwordConfirmation.text,
+                                            nickname = name.text,
+                                            humanVerificationCode = token,
+                                        )
+                                        navigateHome()
+                                    }.onFailure {
+                                      errorMessage = "Are you even a human?"
```
Comment (newmskywalker) [inline android/features/login/src/main/java/io/newm/feature/login/screen/createaccount/CreateAccountScreenPresenter.kt:117]: same here

File: android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt [hunk 5]
```diff
@@ -56,8 +60,12 @@ class ResetPasswordScreenPresenter(
                             isLoading = true
                             coroutineScope.launch {
                                 try {
-                                    signupUseCase.requestEmailConfirmationCode(email.text)
-                                    step = ResetPasswordStep.EnterVerificationCode
+                                    recaptchaClientProvider.get().execute(RecaptchaAction.custom("RequestCode")).onSuccess { token ->
+                                        signupUseCase.requestEmailConfirmationCode(email.text, humanVerificationCode = "")
```
Comment (newmskywalker) [inline android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt]: is it intentional that we're not using the token here?

File: android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt [hunk 6]
```diff
@@ -104,6 +112,7 @@ class ResetPasswordScreenPresenter(
                                             code = authCode.text,
                                             newPassword = password.text,
                                             confirmPassword = passwordConfirmation.text,
+                                            humanVerificationCode = ""
```
Comment (newmskywalker) [inline android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt]: do we need to wrap this with the captcha service? or is that not needed?

**Review by newmskywalker — APPROVED** [review]

File: android/app-newm/src/main/java/io/newm/LoginActivity.kt (hunk 2 above)
Comment (cristhianescobar) [reply android/app-newm/src/main/java/io/newm/LoginActivity.kt:135]: We need to add Sentry to report crashes. That's my next task. However, we might consider doing that, although I think we should restart the app if the user encounters a failure here.

File: android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt (hunk 5 above)
Comment (cristhianescobar) [reply android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt]: Good catch, this needs to send the token as well

File: android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt (hunk 6 above)
Comment (cristhianescobar) [reply android/features/login/src/main/java/io/newm/feature/login/screen/resetpassword/ResetPasswordScreenPresenter.kt]: Same thanks for the catch.

File: shared/src/commonMain/kotlin/io.newm.shared/config/NewmSharedBuildConfig.kt (hunk 1 above)
Comment (martyu) [reply shared/src/commonMain/kotlin/io.newm.shared/config/NewmSharedBuildConfig.kt]: How did you verify they're available at compile time in iOS as an environment variable?  That's how they're used.

Comment (cristhianescobar) [reply shared/src/commonMain/kotlin/io.newm.shared/config/NewmSharedBuildConfig.kt]: I ran the iOS app on Android studio and it worked fine. You don't have to use the same variables if you don't like, you can continue using the secrets file for now.

**Pull request merged**